pkg/ipam/allocator/multipool/pool_allocator.go

```diff
@@ -0,0 +1,402 @@
+			if allocErr != nil {
+				err = errors.Join(err,
+					fmt.Errorf("failed to allocate ipv6 address for node %q from pool %q: %w",
+						cn.Name, reqPool.Pool, allocErr))
+			}
+		}
+	}
+	return err
+}
+
+func (p *PoolAllocator) ReleaseNode(nodeName string) error {
+	// Release CIDRs back into pools
+	var err error
+	for poolName, cidrs := range p.nodes[nodeName] {
+		pool, ok := p.pools[poolName]
+		if !ok {
+			err = errors.Join(err,
+				fmt.Errorf("cannot release from non-existing pool: %s", poolName))
+			continue
+		}
+
+		for cidr := range cidrs.v4 {
+			err = errors.Join(err, releaseCIDR(pool.v4, cidr))
+		}
+		for cidr := range cidrs.v6 {
+			err = errors.Join(err, releaseCIDR(pool.v6, cidr))
+		}
+	}
+
+	// Remove bookkeeping for this node
+	delete(p.nodes, nodeName)
+
+	return err
+}
+
+func (p *PoolAllocator) AllocatedPools(targetNode string) (pools []types.IPAMPoolAllocation) {
+	for poolName, cidrs := range p.nodes[targetNode] {
+		v4CIDRs := cidrs.v4.StringSlice()
+		v6CIDRs := cidrs.v6.StringSlice()
+
+		pools = append(pools, types.IPAMPoolAllocation{
+			Pool:  poolName,
+			CIDRs: append(v4CIDRs, v6CIDRs...),
+		})
+	}
+
+	sort.Slice(pools, func(i, j int) bool {
+		return pools[i].Pool < pools[j].Pool
+	})
+
+	return pools
+}
+
+func (p *PoolAllocator) isAllocated(targetNode, sourcePool string, cidr netip.Prefix) bool {
+	var found bool
+	switch {
+	case cidr.Addr().Is4():
+		_, found = p.nodes[targetNode][sourcePool].v4[cidr]
+	case cidr.Addr().Is6():
+		_, found = p.nodes[targetNode][sourcePool].v6[cidr]
+	}
+	return found
+}
+
+func (p *PoolAllocator) markAllocated(targetNode, sourcePool string, cidr netip.Prefix) {
+	pools, ok := p.nodes[targetNode]
+	if !ok {
+		pools = poolToCIDRs{}
+		p.nodes[targetNode] = pools
+	}
+
+	cidrs, ok := pools[sourcePool]
+	if !ok {
+		cidrs = cidrSets{
+			v4: cidrSet{},
+			v6: cidrSet{},
+		}
+		pools[sourcePool] = cidrs
+	}
+
+	switch {
+	case cidr.Addr().Is4():
+		cidrs.v4[cidr] = struct{}{}
+	case cidr.Addr().Is6():
+		cidrs.v6[cidr] = struct{}{}
+	}
+}
+
+func (p *PoolAllocator) markReleased(targetNode, sourcePool string, cidr netip.Prefix) {
+	pools, ok := p.nodes[targetNode]
+	if !ok {
+		return
+	}
+
+	cidrs, ok := pools[sourcePool]
+	if !ok {
+		return
+	}
+
+	switch {
+	case cidr.Addr().Is4():
+		delete(cidrs.v4, cidr)
+	case cidr.Addr().Is6():
+		delete(cidrs.v6, cidr)
+	}
+
+	// remove pool reference if it is now empty
+	if len(cidrs.v4) == 0 && len(cidrs.v6) == 0 {
+		delete(pools, sourcePool)
+	}
+}
+
+func (p *PoolAllocator) allocateCIDRs(targetNode, sourcePool string, family ipam.Family, toAllocate *big.Int) error {
+	pool, ok := p.pools[sourcePool]
+	if !ok {
+		return fmt.Errorf("cannot allocate from non-existing pool: %s", sourcePool)
+	}
+
+	log.WithFields(logrus.Fields{
+		"targetNode": targetNode,
+		"sourcePool": sourcePool,
+		"family":     family,
+		"toAllocate": toAllocate,
+	}).Debug("allocating cidr")
```

Comment:
This log line might be confusing because it will also log for negative `toAllocate`:

```
level=debug msg="allocating cidr" family=ipv4 sourcePool=default subsys=ipam-allocator-clusterpool-v2 targetNode=kind-worker toAllocate=-26
```

but the loop below will be skipped and nothing allocated. Probably better to skip the log message if `toAllocate <= 0`.